library/core/src/ops/index_range.rs

```diff
@@ -0,0 +1,215 @@
+        self.start = unsafe { unchecked_add(value, 1) };
+        value
+    }
+
+    /// # Safety
+    /// - Can only be called when `start < end`, aka when `len > 0`.
+    unsafe fn next_back_unchecked(&mut self) -> usize {
+        // SAFETY: The range isn't empty, so this cannot overflow
+        let value = unsafe { unchecked_sub(self.end, 1) };
+        self.end = value;
+        value
+    }
+
+    /// Removes the first `n` items from this range, returning them as an `IndexRange`.
+    /// If there are fewer than `n`, then the whole range is returned and
+    /// `self` is left empty.
+    ///
+    /// This is designed to help implement `Iterator::advance_by`.
+    #[inline]
+    pub fn take_prefix(&mut self, n: usize) -> Self {
+        let mid = if n <= self.len() {
+            // SAFETY: We just checked that this will be between start and end,
+            // and thus the addition cannot overflow.
+            unsafe { unchecked_add(self.start, n) }
+        } else {
+            self.end
+        };
+        let prefix = Self { start: self.start, end: mid };
+        self.start = mid;
+        prefix
+    }
+
+    /// Removes the last `n` items from this range, returning them as an `IndexRange`.
+    /// If there are fewer than `n`, then the whole range is returned and
+    /// `self` is left empty.
+    ///
+    /// This is designed to help implement `Iterator::advance_back_by`.
+    #[inline]
+    pub fn take_suffix(&mut self, n: usize) -> Self {
+        let mid = if n <= self.len() {
+            // SAFETY: We just checked that this will be between start and end,
+            // and thus the addition cannot overflow.
+            unsafe { unchecked_sub(self.end, n) }
+        } else {
+            self.start
+        };
+        let suffix = Self { start: mid, end: self.end };
+        self.end = mid;
+        suffix
+    }
+}
+
+impl Iterator for IndexRange {
+    type Item = usize;
+
+    #[inline]
+    fn next(&mut self) -> Option<usize> {
+        if self.len() > 0 {
+            // SAFETY: We just checked that the range is non-empty
+            unsafe { Some(self.next_unchecked()) }
+        } else {
+            None
+        }
+    }
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let len = self.len();
+        (len, Some(len))
+    }
+
+    fn advance_by(&mut self, n: usize) -> Result<(), usize> {
+        let original_len = self.len();
+        self.take_prefix(n);
+        if n > original_len { Err(original_len) } else { Ok(()) }
+    }
+
+    fn try_fold<A, F, R>(&mut self, mut accum: A, mut f: F) -> R
+    where
+        Self: Sized,
+        F: FnMut(A, Self::Item) -> R,
+        R: Try<Output = A>,
+    {
+        while self.len() > 0 {
+            // SAFETY: loop condition ensures that `len > 0` here.
+            let x = unsafe { self.next_unchecked() };
+            accum = f(accum, x)?;
+        }
+
+        try { accum }
+    }
```

Comment:
Do these internal iteration overrides actually perform better than the default implementation? `Range` doesn't implement them, and looking at this I wouldn't think `IndexRange` would benefit from them either.

Reply:
array::IntoIter::fold uses ByRefSized which calls try_fold on the range.

Reply:
Yes, but that does not answer my question of whether overriding `fold` and friends on either `Range` or `IndexRange` is helpful. For `array::IntoIter` it's just convenient to forward `next` to `alive.next`, `fold` to `alive.fold`, etc.

FWIW I'm equally surprised by why `array::IntoIter` overrides `fold` if it calls `Range::fold` which is just the default `fold` implementation. I've seen the PR where it was added, but it has no answers. The `vec::bench_flat_map_collect` benchmark doesn't perform better or worse for me when I comment out the `array::IntoIter::fold` implementation. Maybe something else has changed in the mean time?

Reply:
I'm so used to just adding these that I just did, but you're right that they're plausibly not needed -- the `len() > 0` in `next` (and `next_back`) is probably enough to tell LLVM all it needs to know to canonicalize the loop.

> Maybe something else has changed in the mean time?

Well, LLVM15 came in, which can have massive impacts.  And MIR inlining is also recent, so that might have been relevant.

Reply:
Yup, the manual (try_)folds weren't needed, and I've removed them.